### ISSUE TITLE
Azure Germany: Error refreshing token: invalid character '<' looking for beginning of value.

### ISSUE THREAD

**Issue by StefanScherer** [hashicorp/terraform#10666]
I tried terraform with my Azure Germany free trial, but have problems to make it work/login. I tried the same steps in global Azure and location `West Europe` instead of `Germany Central` without a problem.

### Terraform Version
```
$ terraform -v
Terraform v0.7.13
```

### Affected Resource(s)
Please list the resources as a list, for example:
- azurerm
- azurerm_resource_group

### Terraform Configuration Files
```hcl
variable "location" { default = "Germany Central" }

resource "azurerm_resource_group" "test" {
  name     = "HelloGermany"
  location = "${var.location}"
}
```

### Debug Output
Gist: https://gist.github.com/StefanScherer/5da2a4b3c30f75a8cc6629c1c354cd6c

### Panic Output
If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`.

### Expected Behavior
What should have happened?

It should work like in public Azure:

```
$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but
will not be persisted to local or remote state storage.

azurerm_resource_group.test: Refreshing state... (ID: /subscriptions/guid/resourceGroups/HelloWorld)

The Terraform execution plan has been generated and is shown below.
Resources are shown in alphabetical order for quick scanning. Green resources
will be created (or destroyed and then created if an existing resource
exists), yellow resources are being changed in-place, and red resources
will be destroyed. Cyan entries are data sources to be read.

Note: You didn't specify an "-out" parameter to save this plan, so when
"apply" is called, Terraform can't guarantee this is what will execute.

- azurerm_resource_group.test

+ azurerm_resource_group.test
    location: "germanycentral"
    name:     "HelloGermany"
    tags.%:   "<computed>"


Plan: 1 to add, 0 to change, 1 to destroy.
```
### Actual Behavior
What actually happened?

```
$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but
will not be persisted to local or remote state storage.

Error refreshing state: 1 error(s) occurred:

* Cannot request provider registration for Azure Resource Manager: Error obtaining authorization token: Error refreshing token: invalid character '<' looking for beginning of value.
```

### Steps to Reproduce
1. Create a free trial for Azure Germany https://azure.microsoft.com/en-us/free/germany/
2. Create credentials in portal as described in https://www.terraform.io/docs/providers/azurerm/
3. fetch german locations with `azure location list`, found `Germany Central` and `Germany Northeast`
3. `terraform plan`

### Important Factoids
Azure Germany

### References
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:
- pull request #9765

**Comment by pburkholder** [comment]
@StefanScherer Does this still fail with v0.8.1?  (I'm betting it does, but with a new error).

**Comment by StefanScherer** [comment]
Yes, still failing, both with 0.8.1 and 0.8.2.

I've put Fiddler in between

Login request:
```
GET https://login.microsoftonline.de/ARM_TENANT_ID/OAuth2/authorize?client_id=ARM_CLIENT_ID&resource=https%3a%2f%2fmanagement.azure.com%2f&api-version=1.0&redirect_uri=https%3a%2f%2flogin.microsoftonline.com%2fARM_TENANT_ID%2fNationalCloudFederation&state=SOME_RAW_DATA&response_mode=form_post HTTP/1.1
Host: login.microsoftonline.de
User-Agent: Go-http-client/1.1
Referer: https://login.microsoftonline.com/ARM_TENANT_ID/oauth2/token?api-version=1.0
Accept-Encoding: gzip
Connection: close
```

Login response is a HTML page:

```
HTTP/1.1 200 OK
Cache-Control: no-cache, no-store
Pragma: no-cache
Content-Type: text/html; charset=utf-8
Expires: -1
Vary: Accept-Encoding
Server: Microsoft-IIS/8.5
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
x-ms-request-id: SOME_UUID
P3P: CP="DSP CUR OTPi IND OTRi ONL FIN"
Set-Cookie: buid=RAW_DATA
Set-Cookie: esctx=RAW_DATA
X-Powered-By: ASP.NET
Date: Thu, 22 Dec 2016 06:40:08 GMT
Connection: close
...
```

I think the problem are the ressource=management.azure.com and resource_uri=login.microsoftonline.com as the german azure cloud does not communicate with the public azure cloud.

**Comment by StefanScherer** [comment]
Thanks @pburkholder for https://github.com/Azure/azure-cli/pull/1598

This should be the information needed to fix terraform for other Azure US Gov, Germany, China cloud.

https://github.com/pburkholder/azure-cli/blob/376917c2c306cd577d2f7a3dea9349a7d6124284/src/azure-cli-core/azure/cli/core/cloud.py#L120-L147

Check for trailing slashes, I couldn't test it with `az` cli yet but I remember I had problems and switched to older `azure` cli.

**Comment by StefanScherer** [comment]
The endpoints can be found in the vendored Azure/go-autorest https://github.com/hashicorp/terraform/blob/351c6bed79abbb40e461d3f7d49fe4cf20bced41/vendor/github.com/Azure/go-autorest/autorest/azure/environments.go#L99

**Comment by StefanScherer** [comment]
Seems like `EnvironmentFromName` should be called to get all endpoints.

**Comment by StefanScherer** [comment]
But there also is https://github.com/hashicorp/terraform/blob/fc6e2a1b790f8a2d875c2afc74899eb18ef81b23/vendor/github.com/jen20/riviera/azure/endpoints.go#L16
Don't know the code base enough to understand quickly which one is used, or both.

**Comment by pburkholder** [comment]
HI @StefanScherer,

Thanks for the updates! 

I'm not sure where to start in on this either -- it's clear that Terraform azurerm simply simply doesn't work w/ the current end point configuration. I hope we can get some guidance from Hashicorp as my colleague, @glbrtmrgn, and I would be happy to work on the PR if we can get some pointers.

Note that at https://github.com/hashicorp/terraform/blob/master/builtin/providers/azurerm/provider.go#L121 theres' support for a `ManagementURL` but it's never used as was pointed out last year: https://github.com/hashicorp/terraform/pull/4226#r47362510 

in short, it seems the provider configuration, https://github.com/hashicorp/terraform/blob/master/builtin/providers/azurerm/provider.go#L19-L46
needs to support something like a `management_url` or better `azure_cloud` with values being one of `AzureCloud` or `AzureUSGovernment` or `AzureGermany` or `AzureChina`.
